linux/bindings/src/thread.rs

```diff
@@ -0,0 +1,271 @@
+// SPDX-License-Identifier: GPL-2.0
+// Copyright (c) 2023 Ant Group CO., Ltd.
+
+//! Kernel thread.
+
+use core::{
+    marker::PhantomData,
+    ptr::NonNull,
+    sync::atomic::{AtomicBool, Ordering},
+    time::Duration,
+};
+use kernel::{prelude::*, types::Opaque};
+
+/// A wrapper for the kernel's `struct task_struct`.
+///
+/// Use `Opaque` type to wrap the `struct task_struct`, which is meant to be
+/// used with FFI objects that are never interpreted by Rust code.
+#[repr(transparent)]
+pub struct Thread(Opaque<crate::task_struct>);
+
+/// A struct to carry the user closure and data for `Thread`.
+struct ThreadItem<T> {
+    func: Option<Box<dyn FnOnce() -> T>>,
+    data: Option<T>,
+    is_finished: AtomicBool,
+}
+
+impl<T> ThreadItem<T> {
+    /// Constructs a `ThreadItem`.
+    pub fn try_new<F>(func: F) -> Result<Box<Self>>
+    where
+        F: FnOnce() -> T,
+        F: Send + 'static,
+        T: Send + 'static,
+    {
+        let func = Box::try_new(func)?;
+        Ok(Box::try_new(Self {
+            func: Some(func),
+            data: None,
+            is_finished: AtomicBool::new(false),
+        })?)
+    }
+
+    /// Executes the user closure.
+    ///
+    /// # Panics
+    ///
+    /// This method must only be called once.
+    pub fn execute(&mut self) {
+        let closure = self.func.take().expect("take closure more than once");
+        self.data = Some(closure());
+        self.is_finished.store(true, Ordering::Release);
+    }
+
+    /// Retrieves the data, produced by the closure.
+    pub fn data(&mut self) -> Option<T> {
+        self.data.take()
+    }
+}
+
+impl Thread {
+    /// Gets a reference from a raw pointer.
+    ///
+    /// # Safety
+    ///
+    /// The caller must ensure that the provided raw pointer is not dangling, and
+    /// points at a valid task_struct.
+    pub unsafe fn from_raw<'a>(ptr: *mut crate::task_struct) -> &'a Thread {
+        &*(ptr as *const Thread)
```

Comment:
This is a hugely unsafe function. The safety section fails to mention that `&'a` must be a valid lifetime. Having this method is probably not a good idea. It seems that you prefer to work with `*mut crate::task_struct`. Instead of `Thread`.